flink-connector-gcp-pubsub/src/main/java/org/apache/flink/connector/gcp/pubsub/source/reader/PubSubSplitReader.java

```diff
@@ -0,0 +1,185 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.flink.connector.gcp.pubsub.source.reader;
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.connector.base.source.reader.RecordsBySplits;
+import org.apache.flink.connector.base.source.reader.RecordsWithSplitIds;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitReader;
+import org.apache.flink.connector.base.source.reader.splitreader.SplitsChange;
+import org.apache.flink.connector.gcp.pubsub.source.split.PubSubSplit;
+import org.apache.flink.streaming.connectors.gcp.pubsub.common.PubSubDeserializationSchema;
+import org.apache.flink.streaming.connectors.gcp.pubsub.common.PubSubSubscriber;
+import org.apache.flink.streaming.connectors.gcp.pubsub.common.PubSubSubscriberFactory;
+import org.apache.flink.util.Collector;
+
+import com.google.auth.Credentials;
+import com.google.pubsub.v1.ReceivedMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * A {@link SplitReader} to read from a given {@link PubSubSubscriber}.
+ *
+ * @param <T> the type of the record.
+ */
+public class PubSubSplitReader<T> implements SplitReader<Tuple2<T, Long>, PubSubSplit> {
+    private static final Logger LOG = LoggerFactory.getLogger(PubSubSplitReader.class);
+    private static final long UPCOMING_CHECKPOINT = 0;
+    private final PubSubDeserializationSchema<T> deserializationSchema;
+    private final PubSubSubscriberFactory pubSubSubscriberFactory;
+    private final Credentials credentials;
+    private PubSubSubscriber subscriber;
+    private final PubSubCollector collector;
+    // Store the IDs of GCP Pub/Sub messages that yet have to be acknowledged so that they are not
+    // resent. Must be synchronized because it's accessed both by the fetcher and the reader thread.
+    private final SortedMap<Long, List<String>> messageIdsToAcknowledge =
+            Collections.synchronizedSortedMap(new TreeMap<>());
+
+    /**
+     * @param deserializationSchema a deserialization schema to apply to incoming message payloads.
+     * @param pubSubSubscriberFactory a factory from which a new subscriber can be created from
+     * @param credentials the credentials to use for creating a new subscriber
+     */
+    public PubSubSplitReader(
+            PubSubDeserializationSchema deserializationSchema,
```

Comment:
Can this be changed to `PubSubDeserializationSchema<T>` so we avoid assigning a raw type to a generic type? The code that calls the constructor seems to properly pass the `OUT` type parameter around.